projects/cm_mcu/clocksynth.c

```diff
@@ -153,3 +154,92 @@ int load_clock()
   status = write_register(PostambleList, row);
   return status;
 }
+
+#ifdef REV2
+// return the string that corresponds to the programmed file. If
+// there is an error, an empty string is returned.
+void getClockProgram(int device, char progname_clkdesgid[CLOCK_PROGNAME_REG_NAME], char progname_eeprom[CLOCK_EEPROM_PROGNAME_REG_NAME])
+{
+  // first clear out the return value
+  memset(progname_clkdesgid, '\0', CLOCK_PROGNAME_REG_NAME);
+  memset(progname_eeprom, '\0', CLOCK_EEPROM_PROGNAME_REG_NAME);
+  // ensure that the device is in the right range 0-5
+  if (device < 0 || device > 4)
+    return;
+
+  // grab the semaphore to ensure unique access to I2C controller
+  if (acquireI2CSemaphore(i2c2_sem) == pdFAIL) {
+    log_warn(LOG_SERVICE, "could not get semaphore in time\r\n");
+    return;
+  }
+  // extract info about device
+  uint8_t mux_addr, mux_bit, dev_addr;
+  uint16_t eeprom_progname_reg;
+  // In Rev2 device 0 and devices 1-5 are different and hence are stored in different arrays
+  // for monitoring purposes
+  if (device == 0) {
+    mux_addr = clkr0a_moni2c_addrs[0].mux_addr;
+    mux_bit = clkr0a_moni2c_addrs[0].mux_bit;
+    dev_addr = clkr0a_moni2c_addrs[0].dev_addr;
+    eeprom_progname_reg = clkr0a_moni2c_addrs[0].eeprom_progname_reg;
+  }
+  else {
+    mux_addr = clk_moni2c_addrs[device - 1].mux_addr;
+    mux_bit = clk_moni2c_addrs[device - 1].mux_bit;
+    dev_addr = clk_moni2c_addrs[device - 1].dev_addr;
+    eeprom_progname_reg = clk_moni2c_addrs[device - 1].eeprom_progname_reg;
+  }
+  // set mux bit
+  int status = apollo_i2c_ctl_w(CLOCK_I2C_DEV, mux_addr, 1, 1 << mux_bit);
+  if (status != 0) {
+    log_debug(LOG_I2C, "mux write stat=%d\r\n", status); // can't return due to semaphore
+  }
+  else {
+    // set the page
+    uint8_t page = (CLOCK_PROGNAME_REG_ADDR_START >> 8) & 0xFF;
+    status = apollo_i2c_ctl_reg_w(CLOCK_I2C_DEV, dev_addr, 1,
+                                  CLOCK_CHANGEPAGE_REG_ADDR, 1, page);
+
+    // now read out the six bytes of data in two reads
+    const uint8_t reg = (CLOCK_PROGNAME_REG_ADDR_START)&0xFF;
+    uint16_t init_postamble_page = 32 * (device + 1) - 1;
+
+    uint32_t PreambleList_row; // the size of preamble list in a clock config file store at the end of the last eeprom page of a clock
```

Comment:
can you add some comments here what this is doing? and pls keep the lines within 80 comments

Reply:
got it. I will add a descriptive comment for this.

Reply:
Done